x/wasm-storage/keeper/filter.go

```diff
@@ -39,3 +60,55 @@ func ApplyFilter(filter []byte, reveals []RevealBody) ([]int, bool, error) {
 		return outliers, false, nil
 	}
 }
+
+func filterMode(dataPath string, exitCodes []uint8, reveals []string) ([]int, bool) {
+	var nonZeroExitCode int
+	vals := make([]string, 0, len(reveals))
+
+	for i, r := range reveals {
+		vals = append(vals, gjson.Get(r, dataPath).String())
+		if exitCodes[i] != 0 {
+			nonZeroExitCode++
+		}
+	}
+	outliers, consensus := calcOutliers(vals, exitCodes)
+
+	// If more than 1/3 reveals have non exit code, we reached a consensus that
+	// these are unusable reveals.
+	if consensus || nonZeroExitCode*3 > len(reveals) {
+		return outliers, true
+	}
+	return outliers, false
+}
+
+func calcOutliers[T comparable](reveals []T, exitCode []uint8) ([]int, bool) {
+	freq := make(map[T]int, len(reveals))
+	outliers := make([]int, len(reveals))
+	for i := 0; i < len(outliers); i++ {
+		outliers[i] = 1
+	}
+
+	var maxFreq int
+
+	for i, r := range reveals {
+		if exitCode[i] != 0 {
+			continue
+		}
```

Comment:
shouldn't the exit code of a reveal treated the same way as the reveal (`r` here)?

Reply:
not sure what you mean, but I have some code refactored. reopen this if you still want changes.